### PR TITLE
Add created_at, created_by, updated_at, updated_by to OrganizationMember

### PR DESCRIPTION
**Pull request by tdrivas**
Following the ProjectCollaborator model, the following changed have been made:

- Added created_at and updated_at fields, which capture the timestamps for when the record is created or updated.
- Added created_by and updated_by fields, that track which user created or last updated an OrganizationMember instance.

---

 - [x] Make the `created_by` and `updated_by` to be readonly and automatically set to the current admin user on create.
 - [x] Remove the signal. The `created_by` it should not be updated on delete the user of an organization member.
 - [x] Remove dedicated organization member admin page.
 - [x] Check if the organization member endpoint works https://app.qfield.cloud/swagger/#/api/api_v1_members_create
 . ...

### CODE REVIEW

**Comment by duke-nyuki** [comment]
Task linked: [QF-4535 Add created_at, created_by, updated_at, updated_by to OrganizationMember](https://app.clickup.com/t/2192114/QF-4535)

**Review by suricactus — CHANGES_REQUESTED** [review]
I am afraid this PR has changes that are not relevant to the task itself? Can you please check the changes again and force push in a single commit only the relevant changes?

I also see a missing migration. I guess admin might need some updates too.

**Comment by tdrivas** [comment]
1. Admin has been updated
2. Revert initial commit and commited only the new branch changes
3. Migrations fail on serializers.py. Specifically in 
` versions = serializers.Field(child=FileVersionSerializer())`
which i changed to 
`versions = serializers.ListField(child=FileVersionSerializer())`
4. Migration file has been created

File: docker-app/qfieldcloud/core/models.py [hunk 1]
```diff
@@ -740,6 +740,25 @@ class Meta:
...
+        on_delete=models.SET_NULL,
+        related_name="created_organization_members",
+        null=True,
+        blank=True,
+        limit_choices_to=models.Q(type=User.Type.PERSON),
+    )
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_by = models.ForeignKey(
+        User,
+        on_delete=models.SET_NULL,
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/models.py:754]: What we can do is when the user is being deleted, the fields are updated to point to the project owner (if it is a Person) or the organization owner (if it is an Organization). We can achieve that with signals.

File: docker-app/qfieldcloud/core/serializers.py [hunk 2]
```diff
@@ -171,6 +173,7 @@ class OrganizationSerializer(serializers.ModelSerializer):
     membership_is_public = serializers.BooleanField(
         read_only=True, source="membership_role_is_public"
     )
+    teams = serializers.SerializerMethodField()
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/serializers.py:176]: How this is relevant to this PR?

File: docker-app/qfieldcloud/core/serializers.py [hunk 3]
```diff
@@ -517,7 +522,7 @@ class FileVersionSerializer(serializers.Serializer):
 class FileSerializer(serializers.Serializer):
     """NOTE not used for actual serialization, but for documentation suing Django Spectacular."""
 
-    versions = serializers.ListField(child=FileVersionSerializer())
+    versions = serializers.Field(child=FileVersionSerializer())
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/serializers.py:525]: How this is relevant to this PR?

File: docker-app/qfieldcloud/core/serializers.py [hunk 4]
```diff
@@ -535,3 +540,51 @@ class LatestPackageSerializer(serializers.Serializer):
     package_id = serializers.UUIDField()
     packaged_at = serializers.DateTimeField()
     data_last_updated_at = serializers.DateTimeField()
+
+
+class TeamSerializer(serializers.ModelSerializer):
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/serializers.py:545]: How this is relevant to this PR?

File: docker-app/qfieldcloud/core/tests/test_team.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
+import logging
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/tests/test_team.py:1]: How this is relevant to this PR?

File: docker-app/qfieldcloud/core/tests/test_team_member.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
+import logging
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/tests/test_team_member.py:1]: How this is relevant to this PR?

File: docker-app/qfieldcloud/core/urls.py [hunk 7]
```diff
@@ -105,4 +105,24 @@
         deltas_views.ListDeltasByDeltafileView.as_view(),
     ),
     path("deltas/apply/<uuid:projectid>/", deltas_views.ApplyView.as_view()),
+    path(
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/urls.py:108]: Where this comes from? This PR is about adding new fields to a model, not adding new endpoints.

File: docker-app/qfieldcloud/core/views/users_views.py [hunk 8]
```diff
@@ -146,3 +156,244 @@ class ListUserOrganizationsView(generics.ListAPIView):
 
     def get_queryset(self):
         return Organization.objects.of_user(self.request.user)
+
+
+class TeamMemberDeleteViewPermissions(permissions.BasePermission):
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/users_views.py:161]: How this is related to this PR?

File: docker-app/qfieldcloud/urls.py [hunk 9]
```diff
@@ -29,6 +29,7 @@
 )
 from qfieldcloud.authentication import views as auth_views
 from qfieldcloud.core.views import files_views
+
```
Comment (suricactus) [inline docker-app/qfieldcloud/urls.py:32]: Please revert this change, it brings only noise without additional value.

File: docker-app/qfieldcloud/core/admin.py [hunk 10]
```diff
@@ -360,6 +360,29 @@ def has_change_permission(self, request, obj):
         return False
 
 
+@admin.register(OrganizationMember)
+class OrganizationMemberAdmin(admin.ModelAdmin):
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/admin.py:364]: We can discuss whether we require this new admin. IMO we cacn live without it and the adjustments were needed only for `MemberOrganizationInline`.

File: docker-app/qfieldcloud/core/migrations/0079_organizationmember_created_at_and_more.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
+# Generated by Django 4.2.16 on 2024-11-24 07:00
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/migrations/0079_organizationmember_created_at_and_more.py:1]: Please run pre-commit on your PR.

**Comment by suricactus** [comment]
> 1. Admin has been updated

Good, thanks, I am not sure if adding a new Admin view is needed, see my comment. Anyways, keep it for now.

> 2. Revert initial commit and commited only the new branch changes

I am not sure so. Please check your "Files" tab on PR after opening a new pull request. IMO there are a lot of files that are not relevant to this PR, which should be ~100 lines without the migration.

> 3. Migrations fail on serializers.py. Specifically in
>    ` versions = serializers.Field(child=FileVersionSerializer())`
>    which i changed to
>    `versions = serializers.ListField(child=FileVersionSerializer())`

Why is that? What is the error that was happening?

> 4. Migration file has been created

Thanks. But there was no pre-commit run in advance.

File: docker-app/qfieldcloud/urls.py (hunk 9 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/urls.py:32]: Done

File: docker-app/qfieldcloud/core/migrations/0079_organizationmember_created_at_and_more.py (hunk 11 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/migrations/0079_organizationmember_created_at_and_more.py]: It was failing due to permission errors. I skipped it before. Now i fixed the permissions and I did run it. Solved.

File: docker-app/qfieldcloud/core/models.py (hunk 1 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py:754]: into the signals.py to implement something like this:

@receiver(post_delete, sender=User)
def update_organization_member_fields(sender, instance, **kwargs):
     pass
     
 ?

File: docker-app/qfieldcloud/core/views/users_views.py (hunk 8 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/views/users_views.py]: Deleted

File: docker-app/qfieldcloud/core/urls.py (hunk 7 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/urls.py]: Fixed by uploading the original files

File: docker-app/qfieldcloud/core/tests/test_team_member.py (hunk 6 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/tests/test_team_member.py]: Fixed by uploading the original files

File: docker-app/qfieldcloud/core/serializers.py (hunk 4 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/serializers.py]: Fixed by uploading the original files

File: docker-app/qfieldcloud/core/serializers.py (hunk 3 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/serializers.py]: Fixed by uploading the original files

File: docker-app/qfieldcloud/core/serializers.py (hunk 2 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/serializers.py]: Fixed by uploading the original files

File: docker-app/qfieldcloud/urls.py (hunk 9 above)
Comment (suricactus) [reply docker-app/qfieldcloud/urls.py:32]: It is still there.

File: docker-app/qfieldcloud/core/models.py (hunk 1 above)
Comment (suricactus) [reply docker-app/qfieldcloud/core/models.py:766]: yes, correct. try this one

File: docker-app/qfieldcloud/urls.py (hunk 9 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/urls.py]: Done!

File: docker-app/qfieldcloud/core/models.py (hunk 1 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py:766]: it seems to fail as the ownership is not changing. The code for the signals.py:

`@receiver(pre_delete, sender=User)
def update_project_owner_on_member_delete(sender, instance, **kwargs):
    """
    Update project ownership when an organization member is deleted.
    """

    fallback_user = None

    if hasattr(instance, 'organization') and instance.organization:
        fallback_user = instance.organization.organization_owner
    elif hasattr(instance, 'project') and instance.project.owner.type == User.Type.PERSON:
        fallback_user = instance.project.owner

    if fallback_user:
        OrganizationMember.objects.filter(created_by=instance).update(created_by=fallback_user)
        OrganizationMember.objects.filter(updated_by=instance).update(updated_by=fallback_user)`

Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py:768]: Fixed !

File: docker-app/qfieldcloud/core/models.py [hunk 12]
```diff
@@ -823,6 +842,13 @@ def save(self, *args, **kwargs):
     def teamname(self):
         return self.username.replace(f"@{self.team_organization.username}/", "")
 
+    @staticmethod
+    def format_team_name(organization_name, team_name):
```
[the remark this answers is not in the document]
Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py:849]: Do you mean that is it irrelevant with the current PR?

[the remark this answers is not in the document]
Comment (suricactus) [reply docker-app/qfieldcloud/core/models.py:849]: yes, I think so?

File: docker-app/qfieldcloud/core/signals.py [hunk 13]
```diff
@@ -1,8 +1,45 @@
...
+    if fallback_user:
+        OrganizationMember.objects.filter(created_by=instance).update(
+            created_by=fallback_user
+        )
+        OrganizationMember.objects.filter(updated_by=instance).update(
+            updated_by=fallback_user
+        )
+
+    else:
+        print("No fallback user found. Skipping updates.")
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/signals.py:45]: I am not sure what is it about project ownership, so the docs and function name are saying "Update project ownership"? You don't update the owner, you update the created/update by *with* the owner.

Also you are updating all the organization members with the same `fallback_user`, even if the `fallback_user` has nothing to do with the particular project.

"o1" is owned by "u2"
"o2" is owned by "u3"
"u1" is member of "o1"
"u1" is member of "o2"
"o1" is owner of "p1"
"o2" is owner of "p2"
"u1" added member in both "o1" and "o2".

"u1" is deleted

Then "u2" or "u3" will be used for all `OrgranizationMembers` that were added by "u1" in both "o1" and "o2", even though "u2" is not member of "o2" and "u3" is not member of "o1", because of `projects_owned_by_user.first().owner`, or using the `first()`  to be more precise.


Note there is also a print, please use `logging` module whenever needed.

**Review by suricactus — COMMENTED** [review]
I would suggest that we forget about setting the `uploaded_by` and `created_by` on user deletion for now. We can keep them set to null and in the UI we will show: "created by: (deleted user)".

Please remove the signal and confirm the inline admin works.

File: docker-app/qfieldcloud/core/signals.py (hunk 13 above)
Comment (suricactus) [reply docker-app/qfieldcloud/core/signals.py:45]: Btw, having a second look on this, I see two more problems:

1)
```
    projects_owned_by_user = Project.objects.filter(owner=instance)

    if projects_owned_by_user.exists():
        fallback_user = projects_owned_by_user.first().owner
```
If you filter by owner that equals `instance` and then you get the first result owner, then `fallback_user == instance` no?

2) 
```
        try:
            org_membership = OrganizationMember.objects.get(member=instance)
            fallback_user = org_membership.organization.organization_owner

        except OrganizationMember.DoesNotExist:
            raise ValueError(
                f"No OrganizationMember found for user: {instance.username}"
            )
```

What if you have more than one membership? The `get(member=instance)` will fail with `[OrganizationMember.MultipleObjectsReturned](https://docs.djangoproject.com/en/5.1/ref/models/class/#django.db.models.Model.MultipleObjectsReturned) and will raise a `ValueError`.

**Comment by suricactus** [comment]
Ok, summarizing the tasks here for you:

 - [ ] I have checked the admin, please make the `created_by` and `updated_by` to be readonly and automatically set to the current admin user on create. The `created_by` it should not be updated on updating an organization member.
image
 - [ ] Remove the signal
 - [ ] Remove dedicated organization member admin page

File: docker-app/qfieldcloud/core/admin.py [hunk 14]
```diff
@@ -360,24 +360,27 @@ def has_change_permission(self, request, obj):
...
-    def has_direct_delete_permission(self, request, obj):
-        if obj is None:
-            return True
-        return obj.type in (User.Type.PERSON, User.Type.ORGANIZATION)
-
-    def has_change_permission(self, request, obj):
-        if obj is None:
-            return True
-        return obj.type in (User.Type.PERSON, User.Type.ORGANIZATION)
+# @admin.register(OrganizationMember)
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/admin.py:363]: Kill this.

Comment (tdrivas) [reply docker-app/qfieldcloud/core/admin.py]: Killed!

File: docker-app/qfieldcloud/core/admin.py (hunk 10 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/admin.py]: Not needed currently

File: docker-app/qfieldcloud/core/models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py:850]: Correct

File: docker-app/qfieldcloud/core/signals.py (hunk 13 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/signals.py]: As we said, we do not implement signal for now

File: docker-app/qfieldcloud/core/models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (suricactus) [reply docker-app/qfieldcloud/core/models.py:850]: Would you object removing it from the PR then, please?

File: docker-app/qfieldcloud/core/views/members_views.py [hunk 15]
```diff
@@ -106,3 +111,6 @@ def get_object(self):
         return OrganizationMember.objects.get(
             organization=organization_obj, member=member_obj
         )
+
+    def perform_update(self, serializer):
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/members_views.py:115]:
```suggestion
    def perform_update(self, serializer: ...) -> None:
```

Would you mind adding typing whereever possible?

File: docker-app/qfieldcloud/core/admin.py [hunk 16]
```diff
@@ -1305,8 +1305,12 @@ class OrganizationMemberInline(admin.TabularInline):
...
     fk_name = "organization"
     extra = 0
-
-    autocomplete_fields = ("member",)
+    readonly_fields = (
+        "created_by",
+        "created_at",
+        "updated_by",
+        "updated_at",
+    )
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/admin.py:1313]:
```suggestion

    # These fields must be autocomplete due to performance issue in the default Django admin theme, as the foreign key dropdown renders all the options.
    autocomplete_fields = (
        "member",
        "created_by",
        "updated_by",
    )

    readonly_fields = (
        "created_by",
        "created_at",
        "updated_by",
        "updated_at",
    )
```

I remembered

**Review by suricactus — CHANGES_REQUESTED** [review]
I think this would be the last iteration. Thanks for updating the tests too.

File: docker-app/qfieldcloud/core/admin.py (hunk 16 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/admin.py:1318]: Done

File: docker-app/qfieldcloud/core/views/members_views.py (hunk 15 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/views/members_views.py]: Added

File: docker-app/qfieldcloud/core/models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (tdrivas) [reply docker-app/qfieldcloud/core/models.py]: Done! Forgot before to commit the file (I know, have to check always the file changes here :) )

File: docker-app/qfieldcloud/core/admin.py (hunk 16 above)
Comment (suricactus) [reply docker-app/qfieldcloud/core/admin.py:1318]: Can you please add the comment too.

File: docker-app/qfieldcloud/core/views/members_views.py [hunk 17]
```diff
@@ -106,3 +113,6 @@ def get_object(self):
         return OrganizationMember.objects.get(
             organization=organization_obj, member=member_obj
         )
+
+    def perform_update(self, serializer: Serializer) -> None:
```
Comment (suricactus) [inline docker-app/qfieldcloud/core/views/members_views.py:117]:
```suggestion
    def perform_update(self, serializer: OrganizationMemberSerializer) -> None:
```

File: docker-app/qfieldcloud/core/admin.py (hunk 16 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/admin.py:1318]: Sure!

File: docker-app/qfieldcloud/core/views/members_views.py (hunk 17 above)
Comment (tdrivas) [reply docker-app/qfieldcloud/core/views/members_views.py]: Better!

File: docker-app/qfieldcloud/core/views/members_views.py [hunk 18]
```diff
@@ -8,6 +8,7 @@
 from rest_framework.response import Response
 from rest_framework.settings import api_settings
 
+
```
[the remark this answers is not in the document]
Comment (suricactus) [reply docker-app/qfieldcloud/core/views/members_views.py:11]: Adding random new lines in your code makes your diff unnecessary bigger. Avoid this by checking the "Files Changed" tab, identifying such lines and pushing a new commit that addresses the issue.

**Review by suricactus — APPROVED** [review]
Thank you very much. This PR is ready for merge! Congrats! Let's address the changes on the app.qfield.cloud interface before we hit the "Merge" button.

There are two small tasks to be done before being ready:
- see my comment about ghost lines
- please squash all changes in the PR into a single commit.

File: docker-app/qfieldcloud/core/views/members_views.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (tdrivas) [reply docker-app/qfieldcloud/core/views/members_views.py:11]: Ghost was pushed out!

**Comment by tdrivas** [comment]
> Thank you very much. This PR is ready for merge! Congrats! Let's address the changes on the app.qfield.cloud interface before we hit the "Merge" button.
> 
> There are two small tasks to be done before being ready:
> 
> * see my comment about ghost lines
> * please squash all changes in the PR into a single commit.

Done! Next step is to merge this PR with master and afterwards me to update the ```classic``` submodule so to finalize the relevant task?

**Review by suricactus — APPROVED** [review]
Yes, correct.

**Comment by suricactus** [comment]
@tdrivas you will have the pleasure to hit the "Merge pull request" button. :)
